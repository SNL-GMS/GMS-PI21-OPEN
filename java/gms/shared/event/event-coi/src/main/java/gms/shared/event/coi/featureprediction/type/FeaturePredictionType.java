package gms.shared.event.coi.featureprediction.type;

import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import gms.shared.event.coi.featureprediction.value.ArrivalTimeFeaturePredictionValue;
import gms.shared.event.coi.featureprediction.value.FeaturePredictionValue;
import gms.shared.event.coi.featureprediction.value.NumericFeaturePredictionValue;

import java.io.IOException;
import java.util.Map;
import java.util.Optional;

/**
 * Class for an instantiable object that represents the type of feature prediction. Each "type"
 * is associated with a FeaturePredictionValue class; together, they fully define what each type of
 * feature prediction looks like.
 *
 * @param <T> The FeaturePredictionValue class that should be coupled with this type.
 */
@JsonDeserialize(using = FeaturePredictionType.Deserializer.class)
public abstract class FeaturePredictionType<T extends FeaturePredictionValue<?, ?, ?>> {

  public enum TypeNames {
    ARRIVAL_TIME,
    SLOWNESS,
    EMERGENCE_ANGLE,
    RECEIVER_TO_SOURCE_AZIMUTH,
    SOURCE_TO_RECEIVER_AZIMUTH,
    SOURCE_TO_RECEIVER_DISTANCE,
  }

  private final TypeNames name;

  private FeaturePredictionType(
    TypeNames name
  ) {
    this.name = name;
  }

  @JsonValue
  public TypeNames getName() {
    return name;
  }

  @Override
  public String toString() {
    return "FeaturePredictionType{" +
      "name='" + name + '\'' +
      '}';
  }

  public abstract Class<T> getTypeValueClass();

  //
  // Constants for the supported feature prediction types.
  //

  public static final FeaturePredictionType<ArrivalTimeFeaturePredictionValue> ARRIVAL_TIME_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.ARRIVAL_TIME) {
      @Override
      public Class<ArrivalTimeFeaturePredictionValue> getTypeValueClass() {
        return ArrivalTimeFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SLOWNESS_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SLOWNESS) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> EMERGENCE_ANGLE_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.EMERGENCE_ANGLE) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> RECEIVER_TO_SOURCE_AZIMUTH_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.RECEIVER_TO_SOURCE_AZIMUTH) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SOURCE_TO_RECEIVER_AZIMUTH_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SOURCE_TO_RECEIVER_AZIMUTH) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SOURCE_TO_RECEIVER_DISTANCE_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SOURCE_TO_RECEIVER_DISTANCE) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  /**
   * Custom deserializer used by Jackson. The serialized JSON should contain a single field with
   * a single string that is the name of one of the types, and that string will be returned
   * be JsonParser::getValueAsString.
   */
  static class Deserializer extends StdDeserializer<FeaturePredictionType<?>> {

    private static final Map<TypeNames, FeaturePredictionType<?>> nameTypeMap = Map.of(
      TypeNames.ARRIVAL_TIME, ARRIVAL_TIME_PREDICTION_TYPE,
      TypeNames.SLOWNESS, SLOWNESS_PREDICTION_TYPE,
      TypeNames.EMERGENCE_ANGLE, EMERGENCE_ANGLE_PREDICTION_TYPE,
      TypeNames.RECEIVER_TO_SOURCE_AZIMUTH, RECEIVER_TO_SOURCE_AZIMUTH_PREDICTION_TYPE,
      TypeNames.SOURCE_TO_RECEIVER_AZIMUTH, SOURCE_TO_RECEIVER_AZIMUTH_PREDICTION_TYPE,
      TypeNames.SOURCE_TO_RECEIVER_DISTANCE, SOURCE_TO_RECEIVER_DISTANCE_PREDICTION_TYPE
    );

    public Deserializer() {
      this(null);
    }

    public Deserializer(Class<?> vc) {
      super(vc);
    }

    @Override
    public FeaturePredictionType<?> deserialize(JsonParser p, DeserializationContext ctxt)
      throws IOException {
      return Optional.ofNullable(nameTypeMap.get(TypeNames.valueOf(p.getValueAsString())))
        .orElseThrow();
    }
  }
}
