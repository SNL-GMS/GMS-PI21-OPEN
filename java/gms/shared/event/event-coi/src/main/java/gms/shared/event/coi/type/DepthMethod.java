package gms.shared.event.coi.type;

import com.google.common.base.Functions;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;

import java.util.Arrays;

public enum DepthMethod {
  A("a"),
  F("f"),
  D("d"),
  R("r"),
  G("g"),
  UNKNOWN("-");


  private static final ImmutableMap<String, DepthMethod> LABELS_BY_MODEL;

  static {
    LABELS_BY_MODEL = Arrays.stream(values())
      .collect(ImmutableMap.toImmutableMap(DepthMethod::getLabel, Functions.identity()));
  }

  public final String label;


  DepthMethod(String label) {
    this.label = label;
  }

  public static DepthMethod valueOfLabel(String label) {
    Preconditions.checkNotNull(label, "label cannot be null");
    Preconditions.checkArgument(!label.isBlank(), "label cannot be blank");
    return LABELS_BY_MODEL.get(label);
  }

  public String getLabel() {
    return label;
  }

}

