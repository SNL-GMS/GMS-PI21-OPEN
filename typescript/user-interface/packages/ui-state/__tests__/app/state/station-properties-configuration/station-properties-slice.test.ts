import { createAction } from '@reduxjs/toolkit';
import type * as Redux from 'redux';

import {
  stationPropertiesConfigurationInitialState,
  stationPropertiesConfigurationSlice
} from '../../../../src/ts/app/state/station-properties-configuration/station-properties-slice';
import type { StationPropertiesConfigurationState } from '../../../../src/ts/app/state/station-properties-configuration/types';

describe('station properties configuration slice', () => {
  it('defined', () => {
    expect(stationPropertiesConfigurationInitialState).toBeDefined();
    expect(stationPropertiesConfigurationSlice).toBeDefined();
  });

  it('should return the initial state', () => {
    expect(
      stationPropertiesConfigurationSlice.reducer(undefined, createAction(undefined))
    ).toMatchSnapshot();
    expect(
      stationPropertiesConfigurationSlice.reducer(undefined, createAction(''))
    ).toMatchSnapshot();
    expect(
      stationPropertiesConfigurationSlice.reducer(
        stationPropertiesConfigurationInitialState,
        createAction(undefined)
      )
    ).toMatchSnapshot();
    expect(
      stationPropertiesConfigurationSlice.reducer(
        stationPropertiesConfigurationInitialState,
        createAction('')
      )
    ).toMatchSnapshot();
  });

  it('should update channel configuration columns', () => {
    const action: Redux.AnyAction = {
      type: stationPropertiesConfigurationSlice.actions.updateChannelConfigurationColumns.type,
      payload: {
        channelConfigurationColumns: {
          ...stationPropertiesConfigurationInitialState.channelConfigurationColumns,
          name: false
        }
      }
    };
    const expectedState: StationPropertiesConfigurationState = {
      ...stationPropertiesConfigurationInitialState,
      channelConfigurationColumns: action.payload
    };
    expect(
      stationPropertiesConfigurationSlice.reducer(
        stationPropertiesConfigurationInitialState,
        action
      )
    ).toEqual(expectedState);
  });

  it('should update channel group configuration columns', () => {
    const action: Redux.AnyAction = {
      type: stationPropertiesConfigurationSlice.actions.updateChannelGroupConfigurationColumns.type,
      payload: {
        channelGroupConfigurationColumns: {
          ...stationPropertiesConfigurationInitialState.channelGroupConfigurationColumns,
          name: false
        }
      }
    };
    const expectedState: StationPropertiesConfigurationState = {
      ...stationPropertiesConfigurationInitialState,
      channelGroupConfigurationColumns: action.payload
    };
    expect(
      stationPropertiesConfigurationSlice.reducer(
        stationPropertiesConfigurationInitialState,
        action
      )
    ).toEqual(expectedState);
  });
  it('should set the selected effective at', () => {
    const action: Redux.AnyAction = {
      type: stationPropertiesConfigurationSlice.actions.setSelectedEffectiveAt.type,
      payload: 1
    };
    const expectedState: StationPropertiesConfigurationState = {
      ...stationPropertiesConfigurationInitialState,
      selectedEffectiveAtIndex: action.payload
    };
    expect(
      stationPropertiesConfigurationSlice.reducer(
        stationPropertiesConfigurationInitialState,
        action
      )
    ).toEqual(expectedState);
  });
});
