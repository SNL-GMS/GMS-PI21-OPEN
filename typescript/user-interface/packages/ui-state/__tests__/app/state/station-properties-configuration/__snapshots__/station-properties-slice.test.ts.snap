// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`station properties configuration slice should return the initial state 1`] = `
Object {
  "channelConfigurationColumns": Object {
    "calibrationEffectiveAt": true,
    "calibrationFactor": true,
    "calibrationPeriod": true,
    "calibrationResponseId": true,
    "calibrationStandardDeviation": true,
    "calibrationTimeShift": true,
    "channelBandType": false,
    "channelDataType": false,
    "channelInstrumentType": false,
    "channelOrientationCode": false,
    "channelOrientationType": false,
    "depthKm": true,
    "description": true,
    "eastDisplacementKm": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "fapResponseId": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "nominalSampleRateHz": true,
    "northDisplacementKm": true,
    "orientationHorizontalDegrees": true,
    "orientationVerticalDegrees": true,
    "units": true,
    "verticalDisplacementKm": true,
  },
  "channelGroupConfigurationColumns": Object {
    "depthKm": true,
    "description": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "type": true,
  },
  "selectedEffectiveAtIndex": 0,
}
`;

exports[`station properties configuration slice should return the initial state 2`] = `
Object {
  "channelConfigurationColumns": Object {
    "calibrationEffectiveAt": true,
    "calibrationFactor": true,
    "calibrationPeriod": true,
    "calibrationResponseId": true,
    "calibrationStandardDeviation": true,
    "calibrationTimeShift": true,
    "channelBandType": false,
    "channelDataType": false,
    "channelInstrumentType": false,
    "channelOrientationCode": false,
    "channelOrientationType": false,
    "depthKm": true,
    "description": true,
    "eastDisplacementKm": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "fapResponseId": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "nominalSampleRateHz": true,
    "northDisplacementKm": true,
    "orientationHorizontalDegrees": true,
    "orientationVerticalDegrees": true,
    "units": true,
    "verticalDisplacementKm": true,
  },
  "channelGroupConfigurationColumns": Object {
    "depthKm": true,
    "description": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "type": true,
  },
  "selectedEffectiveAtIndex": 0,
}
`;

exports[`station properties configuration slice should return the initial state 3`] = `
Object {
  "channelConfigurationColumns": Object {
    "calibrationEffectiveAt": true,
    "calibrationFactor": true,
    "calibrationPeriod": true,
    "calibrationResponseId": true,
    "calibrationStandardDeviation": true,
    "calibrationTimeShift": true,
    "channelBandType": false,
    "channelDataType": false,
    "channelInstrumentType": false,
    "channelOrientationCode": false,
    "channelOrientationType": false,
    "depthKm": true,
    "description": true,
    "eastDisplacementKm": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "fapResponseId": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "nominalSampleRateHz": true,
    "northDisplacementKm": true,
    "orientationHorizontalDegrees": true,
    "orientationVerticalDegrees": true,
    "units": true,
    "verticalDisplacementKm": true,
  },
  "channelGroupConfigurationColumns": Object {
    "depthKm": true,
    "description": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "type": true,
  },
  "selectedEffectiveAtIndex": 0,
}
`;

exports[`station properties configuration slice should return the initial state 4`] = `
Object {
  "channelConfigurationColumns": Object {
    "calibrationEffectiveAt": true,
    "calibrationFactor": true,
    "calibrationPeriod": true,
    "calibrationResponseId": true,
    "calibrationStandardDeviation": true,
    "calibrationTimeShift": true,
    "channelBandType": false,
    "channelDataType": false,
    "channelInstrumentType": false,
    "channelOrientationCode": false,
    "channelOrientationType": false,
    "depthKm": true,
    "description": true,
    "eastDisplacementKm": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "fapResponseId": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "nominalSampleRateHz": true,
    "northDisplacementKm": true,
    "orientationHorizontalDegrees": true,
    "orientationVerticalDegrees": true,
    "units": true,
    "verticalDisplacementKm": true,
  },
  "channelGroupConfigurationColumns": Object {
    "depthKm": true,
    "description": true,
    "effectiveAt": true,
    "effectiveUntil": true,
    "elevationKm": true,
    "latitudeDegrees": true,
    "longitudeDegrees": true,
    "name": true,
    "type": true,
  },
  "selectedEffectiveAtIndex": 0,
}
`;
